From 94b2fa8d0f5772f98677b9157cced51e5fdc32f4 Mon Sep 17 00:00:00 2001
From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Tue, 3 Nov 2020 19:16:54 +0800
Subject: [PATCH] cullable: Allow culling of compatible effects

Where a compatible effect is one that does not change geometry or
opacity. This avoids the dimmed parent window of an attached modal dialog
from skipping culling. And that ensures both windows agree perfectly on
the culling regions.

Closes: https://gitlab.gnome.org/GNOME/mutter/-/issues/1500

Also it should improve the rendering performance of attached modal dialogs.

https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1539
---
 src/compositor/meta-cullable.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/compositor/meta-cullable.c b/src/compositor/meta-cullable.c
index 6f38c5e..5dee3e3 100644
--- a/src/compositor/meta-cullable.c
+++ b/src/compositor/meta-cullable.c
@@ -29,7 +29,7 @@
 G_DEFINE_INTERFACE (MetaCullable, meta_cullable, CLUTTER_TYPE_ACTOR);
 
 static gboolean
-has_active_effects (ClutterActor *actor)
+has_incompatible_effects (ClutterActor *actor)
 {
   g_autoptr (GList) effects = NULL;
   GList *l;
@@ -38,7 +38,16 @@ has_active_effects (ClutterActor *actor)
   for (l = effects; l != NULL; l = l->next)
     {
       if (clutter_actor_meta_get_enabled (CLUTTER_ACTOR_META (l->data)))
-        return TRUE;
+        {
+          ClutterEffect *effect = CLUTTER_EFFECT (l->data);
+          gboolean might_change_geometry_or_opacity = TRUE;
+
+          if (CLUTTER_IS_DESATURATE_EFFECT (effect) ||
+              CLUTTER_IS_BRIGHTNESS_CONTRAST_EFFECT (effect))
+            might_change_geometry_or_opacity = FALSE;
+
+          return might_change_geometry_or_opacity;
+        }
     }
 
   return FALSE;
@@ -113,7 +122,7 @@ meta_cullable_cull_out_children (MetaCullable   *cullable,
        * as well for the same reason, but omitted for simplicity in the
        * hopes that no-one will do that.
        */
-      if (needs_culling && has_active_effects (child))
+      if (needs_culling && has_incompatible_effects (child))
         needs_culling = FALSE;
 
       if (needs_culling && !meta_cullable_is_untransformed (META_CULLABLE (child)))
-- 
2.26.2

